mikeio/spatial/FM_geometry.py

```diff
@@ -804,47 +810,52 @@ def _find_element_2d(self, coords: np.array):
 
             # step 2: if not, then try second nearest point
             if not element_found:
-                nodes = self._geometry2d.element_table[few_nearest[k, 1]]
+                candidate = few_nearest[k, 1]
+                assert np.isscalar(candidate)
+                nodes = self._geometry2d.element_table[candidate]
                 element_found = self._point_in_polygon(
                     nc[nodes, 0], nc[nodes, 1], coords[k, 0], coords[k, 1]
                 )
                 ids[k] = few_nearest[k, 1]
 
             # step 3: if not, then try with *many* more points
             if not element_found:
-                # many_nearest, _ = self._find_n_nearest_2d_elements(coords[k:k, :], n=10)
                 many_nearest, _ = self._find_n_nearest_2d_elements(
-                    coords[k, :], n=10
-                )  # JAN fix
-                element_table = self._geometry2d.element_table[many_nearest]
-                lid = self._find_element_containing_point_2d(
-                    element_table,
-                    xn,
-                    yn,
-                    coords[k, 0],
-                    coords[k, 1],
+                    coords[k, :], n=self._max_n_elements(10)
                 )
-                ids[k] = (
-                    many_nearest[lid] if lid > 0 else -1
-                )  # TODO -1 is not a good choice, since it is a valid index
+                for p in many_nearest[2:]:  # we have already tried the two first above
+                    nodes = self._geometry2d.element_table[p]
+                    element_found = self._point_in_polygon(
+                        nc[nodes, 0], nc[nodes, 1], coords[k, 0], coords[k, 1]
+                    )
+                    if element_found:
+                        ids[k] = p
+                        break
+
+                if not element_found:
+                    raise OutsideModelDomainError(x=coords[k, 0], y=coords[k, 1])
```

Comment:
add k as info to error message. And even better - loop over all coords first and find out which are outside and provide user a list of all points outside so he/she does not need to call this method many times just to find out that the next point is also outside